tests/test_experimental.py

```diff
@@ -1,14 +1,392 @@
+import functools
+import time
 from copy import deepcopy
 
 import numpy as np
 import pytest
 
 from iterative_ensemble_smoother import ESMDA
+from iterative_ensemble_smoother.esmda_inversion import normalize_alpha
 from iterative_ensemble_smoother.experimental import (
+    AdaptiveESMDA,
     ensemble_smoother_update_step_row_scaling,
+    groupby_indices,
 )
 
 
+@pytest.mark.parametrize("seed", range(25))
+def test_groupby_indices(seed):
+
+    rng = np.random.default_rng(seed)
+    rows = rng.integers(10, 100)
+    columns = rng.integers(2, 9)
+
+    # Create data matrix
+    X = rng.integers(0, 10, size=(rows, columns))
+
+    groups = list(groupby_indices(X))
+    indices = [set(idx) for (_, idx) in groups]
+
+    # Verify that every row is included
+    union_idx = functools.reduce(set.union, indices)
+    assert union_idx == set(range(rows))
+
+    # Verify that no duplicate rows occur
+    intersection_idx = functools.reduce(set.intersection, indices)
+    assert intersection_idx == set()
+
+    # Verify each entry in the groups
+    for (unique_row, indices_of_row) in groups:
+
+        # Repeat this unique row the number of times it occurs in X
+        repeated = np.repeat(
+            unique_row[np.newaxis, :], repeats=len(indices_of_row), axis=0
+        )
+        assert np.allclose(X[indices_of_row, :], repeated)
+
+
+@pytest.fixture()
+def linear_problem(request):
+
+    # Seed the problem using indirect parametrization:
+    # https://docs.pytest.org/en/latest/example/parametrize.html#indirect-parametrization
+    rng = np.random.default_rng(request.param)
+
+    # Create a problem with g(x) = A @ x
+    num_parameters = 50
+    num_observations = 10
+    num_ensemble = 100
+
+    A = np.exp(rng.standard_normal(size=(num_observations, num_parameters)))
+
+    def g(X):
+        """Forward model."""
+        return A @ X
+
+    # Create observations
+    x_true = np.linspace(-1, 1, num=num_parameters)
+    observations = g(x_true) + rng.standard_normal(size=num_observations) / 10
+
+    # Initial ensemble and covariance
+    X = rng.normal(size=(num_parameters, num_ensemble))
+    covariance = rng.triangular(0.1, 1, 1, size=num_observations)
+    yield X, g, observations, covariance, rng
+
+
+class TestAdaptiveESMDA:
+    @pytest.mark.parametrize(
+        "linear_problem",
+        range(25),
+        indirect=True,
```

Comment:
I have not used `indirect=True` before so bear with me.
The way I read this is that we run the test 25 times where before each run `linear_problem` is run with `rng` set as `default_rng(i)`.
What's the benefit of this versus running once with `rng` set to `default_rng()` so that we get a different seed every time the test is run?
I guess we get broader coverage since the test is run with 25 different seeds, but over time using `default_rng()` will have even better coverage since a new seed is used every time.
I may of course just be misunderstanding how this works...

Reply:
As discussed on Slack, I agree that this is a good thing to do but that using `hypothesis` might be a better approach. We do not have to make the switch in this PR though.